website/oauth/models/__init__.py

```diff
@@ -0,0 +1,325 @@
+                client_secret=self.client_secret,
+            )
+
+            response = oauth.fetch_request_token(self.request_token_url)
+
+            session.data['oauth_states'][self.short_name] = {
+                'token': response.get('oauth_token'),
+                'secret': response.get('oauth_token_secret'),
+            }
+
+            url = oauth.authorization_url(self.auth_url_base)
+
+        return url
+
+    @abc.abstractproperty
+    def callback_url(self):
+        """The provider URL to exchange the code for a token"""
+        raise NotImplementedError()
+
+    @abc.abstractproperty
+    def client_id(self):
+        """OAuth Client ID. a/k/a: Application ID"""
+        raise NotImplementedError()
+
+    @abc.abstractproperty
+    def client_secret(self):
+        """OAuth Client Secret. a/k/a: Application Secret, Application Key"""
+        raise NotImplementedError()
+
+    default_scopes = list()
+
+    @abc.abstractproperty
+    def name(self):
+        """Human-readable name of the service. e.g.: ORCiD, GitHub"""
+        raise NotImplementedError()
+
+    @abc.abstractproperty
+    def short_name(self):
+        """Name of the service to be used internally. e.g.: orcid, github"""
+        raise NotImplementedError()
+
+    def auth_callback(self, user):
+        session = get_session()
+
+        try:
+            cached_credentials = session.data['oauth_states'][self.short_name]
+        except KeyError:
+            raise PermissionsError("OAuth flow not recognized.")
+
+        if self._oauth_version == OAUTH1:
+            request_token = request.args.get('oauth_token')
+
+            if cached_credentials.get('token') != request_token:
+                raise PermissionsError("Request token does not match")
+
+            response = OAuth1Session(
+                client_key=self.client_id,
+                client_secret=self.client_secret,
+                resource_owner_key=cached_credentials.get('token'),
+                resource_owner_secret=cached_credentials.get('secret'),
+                verifier=request.args.get('oauth_verifier'),
+            ).fetch_access_token(self.callback_url)
+
+        elif self._oauth_version == OAUTH2:
+            state = request.args.get('state')
+
+            if cached_credentials.get('state') != state:
+                raise PermissionsError("Request token does not match")
+
+            response = OAuth2Session(
+                self.client_id,
+                redirect_uri=web_url_for(
+                    'oauth_callback',
+                    service_name=self.short_name,
+                    _absolute=True
+                ),
+            ).fetch_token(
+                self.callback_url,
+                client_secret=self.client_secret,
+                code=request.args.get('code'),
+            )
+
+        info = self._default_handle_callback(response)
+        info.update(self.handle_callback(response))
+
+        try:
```

Comment:
If `ExternalAccount` records are meant to be unique, it's safer to add a unique index--then at this point you can attempt to insert and catch the duplicate key error. See https://github.com/CenterForOpenScience/osf.io/blob/develop/website/addons/osfstorage/model.py#L124-L132 and https://github.com/CenterForOpenScience/osf.io/blob/develop/website/addons/osfstorage/model.py#L187-L193.